### PR TITLE
Code cleanup, small optimisations and single matrix pcc implementation

### PR DESCRIPTION
**Pull request by Uditgulati**
Removed matrix M allocation as it wasn't being used.

Added reformulated equation above pcc function that is used by it.

Slight change in amask, bmask assignment loop if statements to reduce number of assignments.

Fixed some comments that didn't match with the corresponding operation (operands weren't correct).

Added omp statement above one loop by the end.

Rest of the commits are for adding the pcc_matrix function which calculates pcc for one matrix input and is optimised for one matrix input. Renamed previous pcc_matrix to pcc_matmat.

### CODE REVIEW

File: R/interface.R [hunk 1]
```diff
@@ -2,12 +2,19 @@
...
+    bM <- aM
+  } else {
+    p = as.integer(ncol(bM))
+  }
   res <- .C("R_pcc_matrix", aM = as.double(aM),
                             bM = as.double(bM),
-                            n = as.integer(nrow(aM)), # nInd
-                            m = as.integer(ncol(aM)), # nPhe A
-                            p = as.integer(ncol(bM)), # nPhe B
+                            n, # nInd
```
Comment (DannyArends) [inline R/interface.R:15]: These 3 lines have been unnecessarily changed, why touch lines of code when nothing happens ??? 
Also all as.XX calls should happen in the .C interface, since in R types like double and integer don't exist (and storage classes might be dynamically changed so the statement p = 0, is not gauranteed to be an integer now)

File: src/MPCC.cpp [hunk 2]
```diff
@@ -282,8 +282,11 @@ void initialize(int &m, int &n, int &p, int seed,
 //This function is the implementation of a matrix x matrix algorithm which computes a matrix of PCC values
 //but increases the arithmetic intensity of the naive pairwise vector x vector correlation
 //A is matrix of X vectors and B is transposed matrix of Y vectors:
-//P = [ sum(AB) - (sumA)(sumB)/N] /
-//    sqrt[ ( sumA^2 -(1/N) (sum A/)^2)[ ( sumB^2 - (1/N)(sum B)^2) ]
+//P = [ sum(AB) - (sumA)(sumB)/N ] /
```
Comment (DannyArends) [inline src/MPCC.cpp:285]: Why was this part of the comment changed ? and what does it mean ? the commit message on the change doesn't mention why this change was needed and the comment is less readable now for any user coming after you

File: src/MPCC.cpp [hunk 3]
```diff
@@ -355,26 +359,28 @@ int pcc_matrix(int m, int n, int p,
     #pragma omp parallel for private (i,k)
     for (i=0; i<m; i++) {
       for (k=0; k<n; k++) {
-        amask[ i*n + k ] = 1.0;
         if (CHECKNA(A[i*n+k])) { 
           amask[i*n + k] = 0.0;
           A[i*n + k] = 0.0; // set A to 0.0 for subsequent calculations of PCC terms
         }else{
+          amask[ i*n + k ] = 1.0;
```
Comment (DannyArends) [inline src/MPCC.cpp:366]: This change changes nothing, and will make the code slower, because of the way CPU branch prediction work

while(x){
  a =1
  if(something) a = 0
}

is faster than: 

while(x){
  if(something) a = 0
  else   a =1
}

File: src/MPCC.cpp [hunk 4]
```diff
@@ -355,26 +359,28 @@ int pcc_matrix(int m, int n, int p,
...
+    if(!sameAB) { //only do it when A and B are not same
+      //if element in B is missing, set bmask and B to 0
+      #pragma omp parallel for private (j,k)
+      for (j=0; j<p; j++) {
+        for (k=0; k<n; k++) {
+          if (CHECKNA(B[j*n+k])) { 
+            bmask[j*n + k] = 0.0;
+            B[j*n + k] = 0.0; // set B to 0.0 for subsequent calculations of PCC terms
+          }else{
+            bmask[ j*n + k ] = 1.0;
```
Comment (DannyArends) [inline src/MPCC.cpp:381]: Again, branch prediction slowdown

File: src/MPCC.cpp [hunk 5]
```diff
@@ -468,11 +479,19 @@ int pcc_matrix(int m, int n, int p,
     //Compute and assemble composite terms
 
     //SASB=SA*SB
-    VMUL(m*p,SA,SB,SASB);
-    //N*SASB
+    if(!sameAB)
+      VMUL(m*p,SA,SB,SASB);
+    else {
+      #pragma omp parallel for private(i, j)
```
Comment (DannyArends) [inline src/MPCC.cpp:485]: This code is duplicated two times, why not make it a function? 
Also the code you provide now doesn't benefit from MKL vectorization like the VMUL function

File: src/MPCC.cpp [hunk 6]
```diff
@@ -483,14 +502,31 @@ int pcc_matrix(int m, int n, int p,
...
+        for(int j = 0; j < p; j++) {
+          SBSB[i*m + j] = SB[j*m + i] * SB[j*m + i];
+        }
+    }
     //N(SBB)
-    VMUL(m*p,N,SBB,NSBB);
+    if(!sameAB)
+      VMUL(m*p,N,SBB,NSBB);
+    else {
+      #pragma omp parallel for private(i, j)
```
Comment (DannyArends) [inline src/MPCC.cpp:518]: Here is the duplication !

File: src/MPCC.cpp [hunk 7]
```diff
@@ -468,11 +479,19 @@ int pcc_matrix(int m, int n, int p,
     //Compute and assemble composite terms
 
     //SASB=SA*SB
-    VMUL(m*p,SA,SB,SASB);
-    //N*SASB
+    if(!sameAB)
+      VMUL(m*p,SA,SB,SASB);
+    else {
```
Comment (DannyArends) [inline src/MPCC.cpp:484]: Always use brackets when writing an if else statement !!!
I know this is legal, but it is a very very fragile way of coding

**Review by DannyArends — APPROVED** [review]
Hey Udit,

I accepted your changes, but still have some comments on the code style and quality, 
I'll fix those myself before sending them to JICS, but see the comments below

Danny

File: src/MPCC.cpp (hunk 2 above)
Comment (Uditgulati) [reply src/MPCC.cpp:285]: This change fixes some brackets typos and adds an equation which is mutiplied by N in both numerator and denominator. If you see the code, it actually multiplies with N instead of dividing.

File: src/MPCC.cpp (hunk 5 above)
Comment (Uditgulati) [reply src/MPCC.cpp:485]: There is no difference in performance in the above code as compared to VMUL. I think that the OMP multithreading is equivalent to VMUL implementation. Testing on penguin2 gave nearly equal execution time.

File: src/MPCC.cpp (hunk 6 above)
Comment (Uditgulati) [reply src/MPCC.cpp:518]: Yeah, a function for it would be good.
